### PR TITLE
MOD-87: Keep page from scrolling down on every page load

### PR DESCRIPTION
**Pull request by Undo1**
My second attempt at [GCI-62](https://issues.openmrs.org/browse/GCI-62) | [MOD-87](https://issues.openmrs.org/browse/MOD-87) ([Melange task](http://www.google-melange.com/gci/task/view/google/gci2014/5296652202016768)).

I figured out why just added `autoscroll="true"` to showModuleNewRelease.html wasn't working by attempting to type a decent question into Stack Overflow - apparently the element with `autoscroll="true"` needs to have a `ui-view` attribute. 

I agree that this is a much better approach than the one I took in #22 with the `<script>`'s, this doesn't require jQuery. Thanks for pointing that out!

### CODE REVIEW

File: app/partials/showModuleNewRelease.html [hunk 1]
```diff
@@ -10,6 +10,8 @@
 <div class="row">
 
   <div class="jumbotron upload-jumbotron" ng-controller="ReleaseFileCtrl"
+  ui-view="upload_new_release"
```
Comment (elliottwilliams) [inline app/partials/showModuleNewRelease.html]: Out of interest, does `upload_new_release` mean anything? It's not actually a view id, right?

**Comment by elliottwilliams** [comment]
Yep, this is working for me, too, thanks! See my note in the diff—if there's a better way that we can do this than trying to load a bogus view, I'd like to try that. Or maybe I'm not understanding that line correctly?

File: app/partials/showModuleNewRelease.html (hunk 1 above)
Comment (Undo1) [inline app/partials/showModuleNewRelease.html]: @elliottwilliams, it doesn't. Through trial and error I found that I could only make autoscroll work if the element has a ui-view attribute. Do you know of a way to make autoscroll work without faking view IDs?

**Comment by Undo1** [comment]
@elliottwilliams, see the last commit. Looks like we don't actually have to load a bogus view. I didn't understand what ui-view does, I figured it was just a convenient way to reference an element like a CSS class. Apologies! How does it look now?

**Comment by elliottwilliams** [comment]
This might still cause problems: basically what `ui-view` does is tells the renderer where to inject a new view in the page. So if I tell the ui-router to go to state "foobar", and ui-router knows that "foobar"'s view is at the URL "templates/foobar.html", it'll look for a `ui-view` element that either accepts any views (a ui-view with no name) or one with a matching name.

This doesn't break anything right now since we don't load any states from inside the `show.newRelease` state, but if we do that stray `ui-view` will cause rendering bugs.

**Comment by elliottwilliams** [comment]
What if you tried this approach? I'm not sure how much Angular you're familiar with, but you could pass the name of the state into the scope, inside `ShowModuleCtrl` in showControllers.js. Then, in showModule.html on the `ui-view` line, you could have a ternary expression like this:

```html
<div ui-view="release" autoscroll="stateName == 'show.newRelease' ? true : false"
class="release-view-animate"></div>
```

You'd just need to make sure you put a variable like `stateName` into the scope to make this work.

**Comment by Undo1** [comment]
I've actually never used Angular before, this is my first attempt. Thank you so much for walking me through this!

I just tried using the scope approach, and pushed the results [here](https://github.com/Undo1/openmrs-contrib-modulus-ui/commit/2131ba00f905003384a2d73166413cc24abd1312). I found that I needed to use `{{ }}` to make Angular evaluate the expression correctly. Now when I switch between views, Chrome's inspector shows that the `autoscroll=(true|false)` is being assigned correctly, but the page jumps such that only the top label (the big "Releases" one) is visible (and not the entire upload box).

I suppose this makes sense, but I would have expected the box to be as near the top of the screen as it could possibly be. I'm wondering if the browser is somehow trying to scroll before the upload box is rendered?

**Comment by elliottwilliams** [comment]
Yeah, I am wondering the same thing. It might not work to dynamically change `autoscroll` that way, but thanks for experimenting. Worst comes to worst we can just use your previous way (unconventional use of ui-view but it works!), but I'll keep looking to see if I can figure anything out.

**Comment by Undo1** [comment]
I find it odd that autoscroll only works when the element has a ui-view attribute, but apparently that's how it works. I've marked http://www.google-melange.com/gci/task/view/google/gci2014/5296652202016768 as complete, your decision : )

Thanks again for helping me through this, I feel like I already understand Angular a great deal more than when I started!

**Comment by elliottwilliams** [comment]
Yeah, since autoscroll is actually a property of the [ui-view directive][0] that's why. And you're very welcome!

Your version with the scope approach didn't work because you were changing `$scope.stateName` in the controller of that view, which doesn't get evaluated until _after_ the view is loaded. :)

Here's a solution I found, if you like it you're welcome to use it:

```js

// in ShowModuleCtrl:

$scope.autoscrollToRelease = function () {
    return $state.is('show.newRelease');
}
```

Then remove any references to `$scope.stateName`, and in showModule.html, do something like this:

```html
<div ui-view="release" autoscroll="autoscrollToRelease()" class="release-view-animate"></div>
```

What's going on is that `$state.is` is a function from the ui-router library ([documentation][0]) that returns a boolean whether the passed state name is the page's current state. Since I am wrapping it in a function on the scope, it'll get re-evaluated as the state changes. 

[0]: https://angular-ui.github.io/ui-router/site/#/api/ui.router.state.$state#methods_is

**Comment by Undo1** [comment]
Aaah, I think I'm finally starting to figure out the timeline of a page load. The controller code is actually run client side, not server side like a Rails app (which is what I normally work on). I (in my still-miniscule knowledge of the field) think that this solution is the most elegant yet, and doesn't involve any ui-view trickery like a few of them. 

The timeline also explains why Chrome's inspector was showing the autoscroll attribute as being correct, but it just wasn't right at the correct *time* in the page lifecycle. That clears up a lot of confusion.

I've added those changes - in my testing of it, it seems to work perfectly on Chrome, Safari and FF. I'm learning a lot from this task!

File: app/partials/showModuleNewRelease.html [hunk 2]
```diff
@@ -10,6 +10,8 @@
 <div class="row">
 
   <div class="jumbotron upload-jumbotron" ng-controller="ReleaseFileCtrl"
+  ui-view
+  autoscroll="true"
```
Comment (elliottwilliams) [inline app/partials/showModuleNewRelease.html]: You can remove these two lines now, since it's covered in showModule.html:145

**Comment by elliottwilliams** [comment]
Awesome! Glad you are learning a lot. You can remove those two lines in showModuleNewRelease.html (I think you meant to do this but forgot to :smile:), and I'll accept and merge.

**Comment by elliottwilliams** [comment]
Also, make sure you set your melange task back to "needs review"

**Comment by Undo1** [comment]
That's what I get for experimenting with git add -p :smile:. Corrected. There's a couple small merge conflicts, I can try handling those if you'd like. 

I wanted to thank you again for spending time helping me, I really appreciate it. I've learned a lot through this task, maybe more than on any other task.

**Comment by dkayiwa** [comment]
@elliottwilliams what do you think of this pull request?

**Pull request merged**